src/integration-test/java/com/commercetools/sync/integration/sdk2/externalsource/states/StateSyncIT.java

```diff
@@ -0,0 +1,1285 @@
+            .initial(true)
+            .build();
+    final StateSyncOptions stateSyncOptions =
+        StateSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+            .batchSize(3)
+            .errorCallback(
+                (exception, oldResource, newResource, updateActions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                  errorCallBackExceptions.add(exception.getCause());
+                })
+            .warningCallback(
+                (exception, newResource, oldResource) ->
+                    warningCallBackMessages.add(exception.getMessage()))
+            .build();
+    // test
+    final StateSyncStatistics stateSyncStatistics =
+        new StateSync(stateSyncOptions).sync(List.of(stateCDraft)).toCompletableFuture().join();
+
+    assertThat(stateSyncStatistics).hasValues(1, 0, 0, 1, 0);
+    assertThat(errorCallBackExceptions).isNotEmpty();
+    assertThat(errorCallBackMessages).isNotEmpty();
+    assertThat(errorCallBackMessages.get(0))
+        .isEqualTo(format("StateDraft with key: '%s' has invalid state transitions", keyC));
+    assertThat(warningCallBackMessages).isEmpty();
+  }
+
+  @Test
+  void sync_WithStateWithEmptyTransitionShouldBeResolved_ShouldAddErrorMessage() {
+    final StateDraft stateCDraft =
+        StateDraftBuilder.of()
+            .key(keyC)
+            .type(StateTypeEnum.REVIEW_STATE)
+            .roles(StateRoleEnum.REVIEW_INCLUDED_IN_STATISTICS)
+            .transitions(StateResourceIdentifierBuilder.of().id("").build())
+            .initial(true)
+            .build();
+    final StateSyncOptions stateSyncOptions =
+        StateSyncOptionsBuilder.of(CTP_TARGET_CLIENT)
+            .batchSize(3)
+            .errorCallback(
+                (exception, oldResource, newResource, updateActions) -> {
+                  errorCallBackMessages.add(exception.getMessage());
+                  errorCallBackExceptions.add(exception.getCause());
+                })
+            .warningCallback(
+                (exception, newResource, oldResource) ->
+                    warningCallBackMessages.add(exception.getMessage()))
+            .build();
+    // test
+    final StateServiceImpl stateService = new StateServiceImpl(stateSyncOptions);
+    final StateReferenceResolver stateReferenceResolver =
+        new StateReferenceResolver(stateSyncOptions, stateService);
+
+    final CompletionStage<StateDraft> result =
+        stateReferenceResolver.resolveReferences(stateCDraft);
+    result
+        .exceptionally(
+            exception -> {
+              assertThat(exception.getMessage())
+                  .contains(
+                      format(
+                          "Failed to resolve 'transition' reference on StateDraft with key:'%s",
+                          keyC));
+              return null;
+            })
+        .toCompletableFuture()
+        .join();
+  }
+
+  private State createStateInSource(final StateDraft draft) {
```

Comment:
Maybe you can split the tests using a source - ctp - project and put to /internalsource/states ?

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/1043/commits/cdcd860d7e04c223366495161bfffe5de85fb8e4